ports/zephyr/machine_uart.c

```diff
@@ -86,65 +113,122 @@ static mp_obj_t mp_machine_uart_make_new(const mp_obj_type_t *type, size_t n_arg
 }
 
 static void mp_machine_uart_deinit(machine_uart_obj_t *self) {
-    (void)self;
+    m_tracked_free(self->tx_buffer);
+    m_tracked_free(self->rx_buffer);
```

Comment:
Do you need to deinit the IRQ callback here?  Otherwise if it continues to receive characters and the IRQ fires, it puts data in the ringbuf which is now freed.

Also, I think there's a memory leak if you continuously allocate UART objects but never deinit them.  The GC will reclaim the UART but the ringbuffer memory will never be freed.  To fix that can add a finaliser that calls this deinit function (same as you did with `machine.Timer`).

Reply:
Good points. I'll fix all that.